rust/examples/l2cap_bridge.rs

```diff
@@ -0,0 +1,647 @@
+                            BridgeData::CloseSignal => return,
+                        }
+                    }
+                })
+                .await;
+            }
+            Some(tcp_stream) => {
+                let (tcp_reader, tcp_writer) = tcp_stream.into_split();
+
+                // Do tcp stuff when something happens on the l2cap channel.
+                let handle_l2cap_data_future =
+                    do_tcp_on_l2cap_data(l2cap_to_tcp_rx, tcp_writer, l2cap_channel.clone());
+
+                // Do l2cap stuff when something happens on tcp.
+                let handle_tcp_data_future =
+                    do_l2cap_on_tcp_data(tcp_reader, l2cap_channel.clone(), false);
+
+                let (handle_l2cap_result, handle_tcp_result) = join!(handle_l2cap_data_future, handle_tcp_data_future);
+
+                if let Err(e) = handle_l2cap_result {
+                    println!("!!! Error: {e}");
+                }
+
+                if let Err(e) = handle_tcp_result {
+                    println!("!!! Error: {e}");
+                }
+            }
+        };
+
+        Python::with_gil(|_| {
+            // Must hold GIL at least once while/after dropping for Python heap object to ensure
+            // de-allocation.
+            drop(l2cap_channel);
+        });
+
+        Ok(())
+    }
+}
+
+/// L2CAP CoC client bridge: connects to a BLE device, then waits for an inbound
+/// TCP connection on a specified port number. When a TCP client connects, an
+/// L2CAP CoC channel connection to the BLE device is established, and the data
+/// is bridged in both directions, with flow control.
+/// When the TCP connection is closed by the client, the L2CAP CoC channel is
+/// disconnected, but the connection to the BLE device remains, ready for a new
+/// TCP client to connect.
+/// When the L2CAP CoC channel is closed, XXXX
+mod client_bridge {
+    use crate::{
+        do_l2cap_on_tcp_data, do_tcp_on_l2cap_data, run_future_with_current_task_locals, BridgeData,
+    };
+    use bumble::wrapper::{
+        device::{Connection, Device},
+        hci::HciConstant,
+    };
+    use futures::executor::block_on;
+    use owo_colors::OwoColorize;
+    use pyo3::{PyResult, Python};
+    use std::{net::SocketAddr, sync::Arc};
+    use tokio::{
+        join,
+        net::{TcpListener, TcpStream},
+        sync::{mpsc, Mutex},
+    };
+
+    pub struct Args {
+        pub psm: u16,
+        pub max_credits: u16,
+        pub mtu: u16,
+        pub mps: u16,
+        pub bluetooth_address: String,
+        pub tcp_host: String,
+        pub tcp_port: u16,
+    }
+
+    pub async fn start(args: &Args, device: &mut Device) -> PyResult<()> {
+        println!(
+            "{}",
+            format!("### Connecting to {}...", args.bluetooth_address).yellow()
+        );
+        let mut ble_connection = device
+            .connect(&args.bluetooth_address)
+            .await
+            .expect("failed to connect to peer");
```

Comment:
why panic rather than just bubbling up the error?

Reply:
No good reason. Updated.